added express-prisma-and-tailwindcss article (#1172)
oktadeveloper/okta-blog: _source/_posts/2022-05-13-express-prisma-and-tailwindcss.md

---
layout: blog_post

> title: ""

Reviewer: ```suggestion
title: "How to Build an Express application using Prisma"
```

> author:

Reviewer: ```suggestion
author: nickolas-fisher
```

> by: advocate|contractor

Reviewer: ```suggestion
by: contractor
```

> communities: [devops,security,mobile,.net,java,javascript,go,php,python,ruby]

Reviewer: ```suggestion
communities: [javascript]
```

> description: ""

Reviewer: ```suggestion
description: "Learn how to build a workout tracker application using Express and Prisma and secure it using Okta."
```

> tags: []

Reviewer: ```suggestion
tags: [prisma, node, express]
```

tweets:

> - ""

Reviewer: ```suggestion
- "Learn how to build a workout tracker application using Express and Prisma and secure it using Okta ⚡️"
```

> - ""

Reviewer: ```suggestion

```

> - ""

Reviewer: ```suggestion

```

image:

> type: awareness|conversion

Reviewer: ```suggestion
type: conversion
```

---

> ## Building an Express application using Prisma 

Reviewer: ```suggestion
## How to Build an Express application using Prisma
```

[Express](https://expressjs.com/) is one of the most popular web frameworks available today.  Understanding the tools available when building any web application is critical to being a good developer.

> [Prisma](https://www.prisma.io/) is an ORM for Node.js using TypeScript.  It integrates with many of the most popular databases today including MySQL, SQL Server, SQLite, and MongoDB.  CockroachDB is available in preview mode.  Prisma emphasizes a human readable schema with a type-safe database client.  Prisma also includes other features such as migrations, seed data, and a virtual database browser.  

Reviewer: ```suggestion
[Prisma](https://www.prisma.io/) is an ORM for Node.js using TypeScript.  It integrates with many of the most popular databases today including MySQL, SQL Server, SQLite, and MongoDB.   Prisma emphasizes a human-readable schema with a type-safe database client.  Prisma also includes other features such as migrations, seed data, and a virtual database browser.  
```

> In this project, you will use Prisma to connect to a database server.  You will build a schema to model a workout tracker application.  Then you will create some seed data to begin working with and use Prisma to migrate and seed your database.  You will then create the workout web application using [Pug](https://pugjs.org/api/getting-started.html) and [tailwindcss](https://tailwindcss.com/) on your front end to compliment your express backend.  

Reviewer: Should it say 'to run migrations' instead of 'to migrate'?

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/f2a12ff6f5d11671f2a01589ab9eb13d5258725b

> In this project, you will use Prisma to connect to a database server.  You will build a schema to model a workout tracker application.  Then you will create some seed data to begin working with and use Prisma to migrate and seed your database.  You will then create the workout web application using [Pug](https://pugjs.org/api/getting-started.html) and [tailwindcss](https://tailwindcss.com/) on your front end to compliment your express backend.  

Reviewer: ```suggestion
You will use Prisma to connect to a database server in this project.  You will build a schema to model a workout tracker application.  Then you will create some seed data to begin working with and use Prisma to migrate and seed your database.  You will then create the web application using [Pug](https://pugjs.org/api/getting-started.html) and [Tailwind CSS](https://tailwindcss.com/) to build the application frontend.  
```

> ### Prequresites

Reviewer: ```suggestion
### Prerequisites
```

To continue you will need:

> - An IDE for javascript.  I will use [Visual Studio Code](https://code.visualstudio.com/) but you can use Atom, Webstorm, or any other IDE you wish

Reviewer: ```suggestion
- An IDE for JavaScript.  I will use [Visual Studio Code](https://code.visualstudio.com/) but you can use Atom, Webstorm, or any other IDE you prefer
```

- [Node.js](https://nodejs.org/en/)

> - A database, one of PostgreSQL, MySQL, SQLite, SQL Server, or MongoDB.  I will be using SQL Server

Reviewer: What do you think about using an on-disk database like SQLite? It might be easier to run as compared to SQL Server.

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/6a7fe3c83cf50cda128b7cc8326f539139c5fc77

- An [Okta Developer Account](https://developer.okta.com/) (free forever, to handle your OAuth needs)
- [Okta CLI](https://cli.okta.com)

If you wish to look at the app you can [view it on github](https://github.com/nickolasfisher/Okta_Primsa).

## Create your Okta application

> Create a new directory for your application.  

Reviewer: Should we also ask them to cd into that dir?

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/1f6a5339bac9002095a2ce158a63b30574efbf88

{% include setup/cli.md type="web" loginRedirectUri="http://localhost:3000/authorization-code/callback" %}

## Create your Express Application

Next, you will use the `express-generator` application generator tool to quickly scaffold your application.  Run the following command.

```console
npx express-generator --view=pug
```

Next, you will need to install your packages.  

```console
npm i @prisma/client@3.13.0
npm i dotenv@16.0.0
npm i passport@0.5.2
npm i passport-openidconnect@0.1.1
npm i express-session@1.17
npm i -D tailwindcss@3.0.24
npm i -D prisma@3.13.0
```

> `@prisma/client` is used to access the database from your server code
> `dotenv` reads configuration settings from `.env` files like the one produced by the Okta CLI
> `passport` is a middleware for Node.js that is flexible enough to handle most authentication scenarios, including Okta.  `passport-openidconnect` is a module for passport that lets you authenticate with OpenID Connect

> > `express-session` is required for passport.

Reviewer: Perhaps we can add another statement here explaining why?

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/11f8b50149a7ac8617622dc959c652f1cbc59b53

> > `tailwindcss` is the CSS framework you will use.  If you've never used tailwindcss before you may wonder why it's a development dependency.  This is because tailwindcss will dynamically build your CSS files from your views and configuration.  More on that later

Reviewer: ```suggestion
> `Tailwind CSS` is the CSS framework you will use.  If you've never used Tailwind before, you may wonder why it's a development dependency.  This is because Tailwind will dynamically build your CSS files from your views and configuration.  More on that later
```

> `prisma` is the ORM you are using.  This is a dev dependency because the prisma library handles the migrations, seed data, etc, while the `@prisma/client` package is used in your application at runtime.

> At this point, you should initialize tailwindcss and prisma even though you won't configure to use either quite yet.

Reviewer: Tailwind's branding seems to be `Tailwind CSS` (apart from the logo): https://tailwindcss.com/docs/installation. Would you mind updating that throughout the doc? :)

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/bab312ce9a416d93238a8e46f2385c933e5fdcf5

Start with prisma.  Run the following command.

```console
npx prisma init
```

This command will add a new folder named `prisma` to your application. This also adds a file called `.env` to your project with some default configuration. Replace the contents of that file with the code below.

```bash
# Environment variables declared in this file are automatically made available to Prisma.
# See the documentation for more detail: https://pris.ly/d/prisma-schema#accessing-environment-variables-from-the-schema

# Prisma supports the native connection string format for PostgreSQL, MySQL, SQLite, SQL Server, MongoDB, and CockroachDB (Preview).
# See the documentation for all the connection string options: https://pris.ly/d/connection-strings

DATABASE_URL={yourServersConnectionString}
SEED_USER_NAME={yourOktaUserName}
```

> In this example I used Microsoft's SQL Server because that's what I had installed at the time of writing this application.  But one of the cool things about Prisma and most ORMs is that they are built to support many databases.  The boilerplate code includes comments about how to configure your application for other database servers.  Feel free to use what you feel most comfortable with and use the appropriate connection string.

Reviewer: I think installing and creating an empty SQLite DB is quicker to explain in a paragraph here.

Reviewer: Easier for us to potentially check in the DB into the repo as well so that the demo can be run with minimal build steps.

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/6a7fe3c83cf50cda128b7cc8326f539139c5fc77

The `SEED_USER_NAME` setting is used in your seed data.  You want to make sure that it's the same as the Okta username you sign in to.  This will allow the application to associate your logged-in user with the data that you will seed into your database.  

 Next, update your `package.json` file with the following code.

```JSON
{
  "name": "workout-app",
  "version": "0.0.0",
  "private": true,
  "prisma": {
    "seed": "node prisma/seed.js"
  },
  "scripts": {
    "start": "node ./bin/www"
  },
  "dependencies": {
    "@prisma/client": "^3.13.0",
    "cookie-parser": "~1.4.4",
    "debug": "~2.6.9",
    "dotenv": "^16.0.0",
    "express": "~4.16.1",
    "http-errors": "~1.6.3",
    "morgan": "~1.9.1",
    "passport": "^0.5.2",
    "passport-openidconnect": "^0.1.1",
    "pug": "2.0.0-beta11"
  },
  "devDependencies": {
    "prisma": "^3.13.0",
    "tailwindcss": "^3.0.24"
  }
}
```

> This will add the `prisma.seed` command you will need later.  

Reviewer: Should it be `prisma seed`?

Author: https://github.com/oktadev/okta-blog/pull/1172/commits/4a53d6be0eae0850d06be9cc2bfdd0615cb7d0d8
